src/libstd/num/f64.rs

```diff
@@ -84,6 +84,38 @@ mod cmath {
     }
 }
 
+#[cfg(not(target_os = "sunos"))]
+macro_rules! log_wrapper {
+    ($num:ident, $f:ident) => (
+        unsafe { intrinsics::$f($num) }
+    )
+}
+
+// Illumos requires a wrapper around log, log2, and log10 functions
+// because of non-standard behavior (e.g. log(-n) returns -Inf instead
+// of expected NaN).
+#[cfg(target_os = "sunos")]
+macro_rules! log_wrapper {
+    ($num:ident, $f:ident) => (
+        if $num.is_finite() {
+            if $num > 0.0 {
+                return unsafe { intrinsics::$f($num) }
+            }
+            return if $num == 0.0 {
+                NEG_INFINITY // log(0) = -Inf
+            } else {
+                NAN // log(-ve) = NaN
+            }
+        } else if $num.is_nan() {
+            $num // log(NaN) = NaN
+        } else if $num > 0.0 {
+            $num // log(Inf) = Inf
+        } else {
+            return NAN // log(-Inf) = NaN
+        }
+    )
```

Comment:
Instead of a macro, could this be done with something like:

```rust
fn foo<F: Fn(f64) -> f64>(val: f64, f: F) -> f64 {
    if !cfg!(target_os = "sunos") {
        f(val)
    } else {
         // contents of this macro
    }
}
```